src/routes/transactions/mappers/multisig-transaction.mapper.ts

```diff
@@ -0,0 +1,592 @@
+      this.getFromParam(dataDecoded, '') === transaction.safe ||
+      this.getToParam(dataDecoded, '') === transaction.safe
+    );
+  }
+
+  private isCustomTransaction(
+    value: number,
+    dataSize: number,
+    operation: Operation,
+  ): boolean {
+    return (value > 0 && dataSize > 0) || operation !== 0;
+  }
+
+  private isEtherTransfer(value: number, dataSize: number): boolean {
+    return value > 0 && dataSize === 0;
+  }
+
+  private isSettingsChange(
+    transaction: MultisigTransaction,
+    value: number,
+    dataSize: number,
+  ): boolean {
+    return (
+      value === 0 &&
+      dataSize > 0 &&
+      transaction.safe === transaction.to &&
+      this.SETTINGS_CHANGE_METHODS.includes(transaction.dataDecoded?.method)
+    );
+  }
+
+  private isValidTokenTransfer(transaction: MultisigTransaction): boolean {
+    return (
+      (this.isErc20Transfer(transaction) ||
+        this.isErc721Transfer(transaction)) &&
+      this.checkSenderOrReceiver(transaction)
+    );
+  }
+
+  private async mapTransactionInfo(
+    chainId: string,
+    transaction: MultisigTransaction,
+    safe: Safe,
+  ): Promise<TransactionInfo> {
+    const value = Number(transaction?.value) || 0;
+    const dataSize = transaction.data
+      ? (Buffer.byteLength(transaction.data) - 2) / 2
+      : 0;
+
+    if (this.isCustomTransaction(value, dataSize, transaction.operation)) {
+      return await this.mapCustomTransaction(
+        transaction,
+        value,
+        dataSize,
+        chainId,
+      );
+    }
+
+    if (this.isEtherTransfer(value, dataSize)) {
+      return this.getToEtherTransfer(chainId, transaction, safe);
+    }
+
+    if (this.isSettingsChange(transaction, value, dataSize)) {
+      return this.getSettingsChangeTransaction(chainId, transaction, safe);
+    }
+
+    if (this.isValidTokenTransfer(transaction)) {
+      const token = await this.tokenRepository.getToken(
+        chainId,
+        transaction.to,
+      );
+
+      if (token.type === TokenType.Erc20) {
+        return this.mapErc20Transfer(token, chainId, transaction);
+      }
+
+      if (token.type === TokenType.Erc721) {
+        return this.mapErc721Transfer(token, chainId, transaction);
+      }
+    }
+
+    return this.mapCustomTransaction(transaction, value, dataSize, chainId); // TODO: default case
+  }
+
+  private filterAddressInfo(addressInfo: AddressInfo): AddressInfo {
+    return {
+      ...{ value: addressInfo.value },
+      ...(addressInfo.name !== '' && { name: addressInfo.name }),
+      ...(addressInfo.logoUri && { logoUri: addressInfo.logoUri }),
+    };
+  }
+
+  private async mapCustomTransaction(
+    transaction: MultisigTransaction,
+    value: number,
+    dataSize: number,
+    chainId: string,
+  ): Promise<CustomTransactionInfo> {
+    const toAddressInfo = await this.addressInfoHelper.getOrDefault(
+      chainId,
+      transaction.to,
+    );
+    return {
+      type: 'Custom',
```

Comment:
Do you mean extracting the multisig transaction types to an enum? Or using a constructor/builder with a fixed property value for `type`? Could you elaborate, please? :)

Reply:
You return `CustomTransactionInfo`. Does `CustomTransactionInfo` have `Custom` always as the value of `type`? If yes we don't need to parameterise it no?

Reply:
That's right 👍🏻 But `CustomTransactionInfo` is a JS class, if we want to have a fixed value for one of its properties we'd need a constructor. Maybe this could be a good candidate to use a builder pattern, wdyt?

Reply:
Yes I think a constructor would be a good option here:

```typescript
export class CustomTransactionInfo extends TransactionInfo {
  @ApiProperty()
  type = 'Custom';

  ...

  constructor(
    to: AddressInfo | null,
    dataSize: string,
    value: string,
    methodName: string,
    actionCount: number | null,
    isCancellation: boolean,
  ) {
    super();
    this.to = to;
    this.dataSize = dataSize;
    this.value = value;
    this.methodName = methodName;
    this.actionCount = actionCount;
    this.isCancellation = isCancellation;
  }
}
```

Also I think that instead of having `?` in the route properties we should have `T | null`. At this point we should know if the value is present or not and I think it's more semantically correct. Wdyt?

Reply:
I agree, changed in [de4181b](https://github.com/5afe/safe-client-gateway-nest/pull/197/commits/de4181bab770581f3e869e7eadce9fd43d16e5a9)